### ISSUE TITLE
c++ memory leak if switch-on and repeat are combined

### ISSUE THREAD

**Issue by jeclarke** [kaitai-io/kaitai_struct#494]
If a switch on field is set to repeat then the generated c++ code leaks memory in the case where the found value is not covered by the switch cases.

A ksy file like this:
````yaml
meta:
  id: sample
  endian: be

seq:
  - id: sample_type
    type: u1
  - id: sample_field
    repeat: eos
    type:
      switch-on: sample_type
      cases:
        '1': subsample

types:
  subsample:
    seq:
      - id: subsample_field
        type: u1
````

Generates code like this:
````cpp
void sample_t::_read() {
    m_sample_type = m__io->read_u1();
    m_sample_field = new std::vector<subsample_t*>();
    {
        int i = 0;
        while (!m__io->is_eof()) {
            n_sample_field = true;
            switch (sample_type()) {
            case 1: {
                n_sample_field = false;
                m_sample_field->push_back(new subsample_t(m__io, this, m__root));
                break;
            }
            }
            i++;
        }
    }
}

sample_t::~sample_t() {
    if (!n_sample_field) {
        for (std::vector<subsample_t*>::iterator it = m_sample_field->begin(); it != m_sample_field->end(); ++it) {
            delete *it;
        }
        delete m_sample_field;
    }
}
````

Note that `m_sample_field` is always created with new but only deleted if n_sample_field is false. When a case statement matches all is OK but when it does not the vector is not deleted.

I worked around this by introducing a wrapper type for each case and putting the repeat clause in there. Like this:
````yaml
meta:
  id: sample
  endian: be

seq:
  - id: sample_type
    type: u1
  - id: sample_field
    type:
      switch-on: sample_type
      cases:
        '1': wrapper

types:
  wrapper:
    seq:
      - id: items
        repeat: eos
        type: subsample

  subsample:
    seq:
      - id: subsample_field
        type: u1
````

**Comment by jeclarke** [comment]
This is kaitai-struct-compiler 0.8

**Comment by generalmimon** [comment]
@jeclarke Thanks for the report, you were completely right! It was caused by that the nullability for a single switch type was interpreted as the nullability of the whole array of switch types, which is obviously wrong. A new test [`switch_repeat_expr_invalid.ksy`](https://github.com/kaitai-io/kaitai_struct_tests/blob/2b5bd04ce950691760d503506a1ea7eae5f31ddb/formats/switch_repeat_expr_invalid.ksy#L1) tests this situation. It should be fixed now in the recent compiler version.

While thinking about this situation when none of the `cases` match the value of the `switch-on` in some iteration of a `repeat`ed field, I realized the that the current behavior of the generated C++ code might be unsuitable in some cases. As none of the defined `cases` match, the value of the current item is technically `null`, but due to the fact that C++ doesn't have the concept of `null`, nothing is pushed to the array. It behaves as if iteration has never run. You'll end up with an filtered array with only those items that weren't null (that actually had some value).

If the field has the `size` declared and the default case `_` is not explicitly specified, a raw byte array is pushed to the `_raw_items` array by default (if none of the `cases` is matched). When matching a user type in `cases`, the underlying byte array is also pushed to `_raw_items`. That means that `_raw_items` array will reflect contain all the iterations, even when none of the cases were matched (so the value of the current iteration was `null`), but `items` will be a dense array just filled with the user types.

Just a note that one can't rely on that a user type under `items[i]` _always_ corresponds to the byte array at `_raw_items[i]` (where `i` is the same index).
